### ISSUE TITLE
xfb offset overlapping checks fails when combining block and per-member xfb_offset qualifier

### ISSUE THREAD

**Issue by infapi00** [KhronosGroup/glslang#1535]
So the following test.vert file:

```
#version 450

layout(location=5, xfb_offset=0) out block2 {
   float y1_out;
   layout(xfb_offset = 16) vec4 y2_out;
};

void main() {
   y1_out = 7.0;
   y2_out = vec4(1.0, 0.0, 0.0, 1.0);
   gl_Position = vec4(0.0);
}
```
Will fail with the following error:

> $ glslangValidator -G test-member-offset.vert -H
> test-member-offset.vert
> ERROR: test-member-offset.vert:3: 'xfb_offset' : overlapping offsets at offset 16 in buffer 0
> ERROR: 1 compilation errors.  No code generated.

After checking a little the code, this assert is triggered at ParseHelper.cpp, line 5329, method TParseContext::layoutTypeCheck

What it is happening is that this method adds the range defined by the type/offset on its xfb_buffer as it iterates on the type. So it first add the range for y2_out ([16, 16+size(vec4)]). And then the method is called for the full block2, and it throws an error because part of the range needed for the block2 [0, size(block)] is already marked as used.

So there are two problems here:
1. When checking if there is an overlap, it doesn't differentiate members and it containers
2. TIntermediate::computeTypeXfbSize doesn't take into account that one of the members could have a custom offset. So it returns the minimum size needed for the block without custom offsets.

FWIW, when I was checking if fixing by myself those issues were assumable, I started to find how the xfb offsets are dealt not really consistent with other qualifiers. And I think that it is partly due xfb being underspecified on the spec. Specifically with output blocks. So using ARB_gl_spirv spec as reference, quoting:

>     "If the shader used to record output variables for transform feedback
>     varyings uses the xfb_buffer, xfb_offset, or xfb_stride layout qualifiers,
>     or its SPIR-V entry point declares the *Xfb* Execution Mode, the values
>     specified by TransformFeedbackVaryings are ignored, and the set of
>     variables captured for transform feedback is instead derived from the
>     specified layout qualifiers or SPIR-V decorations: outputs specifying both
>     an *XfbBuffer* and an *Offset* are captured, while outputs not specifying
>     both of these are not captured. Values are captured each time the shader
>     writes to such a decorated object."

Note that the last sentence, points a "decorated object", so that could mean a individual variable or a full block. As nothing else is specified, we would need to rely on GLSL spec. So from GLSL 4.60 spec, section "4.4.2 Output Layout Qualifiers", subsection "Transform Feedback Layout Qualifiers":

> If a block is qualified with xfb_offset, all its members are assigned transform feedback buffer offsets. If a block is not qualified with xfb_offset, any members of that block not qualified with an xfb_offset will not be assigned transform feedback buffer offsets. Only variables and block members that are assigned offsets will be captured (thus, a proper subset of a block can be captured).

Right now glslang is just adding on the SPIR-V the xfb offset (just Offset) if it is sets explicitly on the GLSL shader (so on the example used on top, 0 for the block, 16 for y2_out, no offset for y1_out). And from the spec, that's correct. But this seems somewhat inconsistent to what happens with other qualifiers, like bindings, location, offset (on a ubo/ssbo) etc, were it is expected that SPIR-V would include a explicit value for all variables. Even if on GLSL the driver has rules to automatically set all those if the explicit values are missing, when the driver consumes SPIR-V it is expected to being already provided.

In other words, if a explicit xfb_offset is set for a block, I would expect that the SPIR-V generated included a xfb_offset for each member. But I would also find inconsistent to force glslang to be able to assign automatically those offsets. So, on the case of a block being assigned a xfb offset, I would expect the following behaviour:

1. Raise an error if any of the members lack a xfb_offset. Something like "Setting xfb_offset to a block mark all members to be captured. All members needs a xfb_offset"
2. Optionally add a command line option so glslang would automatically assign it (equivalent to aml, something like --auto-map-xfb-offset)

Having said so, all this is a kind of my opinion based on how this work with other qualifiers, with the feeling that this should be more clear on spec. So perhaps it would be needed to open a spec issue first?

**Comment by infapi00** [comment]
> In other words, if a explicit xfb_offset is set for a block, I would expect that the SPIR-V generated included a xfb_offset for each member. But I would also find inconsistent to force glslang to be able to assign automatically those offsets. So, on the case of a block being assigned a xfb offset, I would expect the following behaviour:
> 
>     Raise an error if any of the members lack a xfb_offset. Something like "Setting xfb_offset to a block mark all members to be captured. All members needs a xfb_offset"
>     Optionally add a command line option so glslang would automatically assign it (equivalent to aml, something like --auto-map-xfb-offset)
> 

While debugging/trying to fix issue #1526 I found the method TParseContext::fixBlockXfbOffsets, that basically do what I call there as an optional --auto-map-xfb-offset. So forget the last part of the issue description, as that already exists, and it is seems that the bug is that sometimes is not being applied.

Having said so, it is likely that is a different problem, that deserves a specific issue. Will investigate a little more before doing it though.

**Comment by johnkslang** [comment]
So, the question is, what is the specified behavior?  I see:

>  If a block is qualified with **xfb_offset**, all its members are assigned transform feedback buffer offsets.

But, I don't see the specification saying you can assign contradictory block and member offsets and that that is okay, with the member overriding the block. Do you see that?

**Comment by infapi00** [comment]
>  But, I don't see the specification saying you can assign contradictory block and member offsets and that that is okay, with the member overriding the block. Do you see that?

Yes. From GLSL 4.60 spec, section 4.4.2 "Output Layout Qualifiers", subsection "Transform Feedback Layout Qualifiers", on the "For example" table:

> layout(xfb_buffer = 3, xfb_offset = 12) out block2 {
>     vec4 v; // v will be written to byte offsets 12 through 27 of buffer
>     float u; // u will be written to offset 28
>     layout(xfb_offset = 40) vec4 w;
>     vec4 x; // x will be written to offset 56, the next available offset
> };

So on that example, w is getting an offset 40, so it xfb_offset is overriding the "automatic" assignment from the block. The thing is that the spec words the block assignment in terms of "next available offset". Slightly before, and somewhat mixed with the doubles explanation, we can find this:

> Then, within the qualified entity, subsequent components are each assigned, in
> order, to the next available offset aligned to a multiple of that component’s size.

So basically, member xfb_offset overriding the block would be ok as far as it fulfills the alignment.

And finally, AFAIU, right now the code is already handling this case. From TParseContext::fixXfbOffsets:

>         // see if we need to auto-assign an offset to this member
>         if (! memberQualifier.hasXfbOffset()) {
>             // "if applied to an aggregate containing a double, the offset must also be a multiple of 8"
>             if (containsDouble)
>                 RoundToPow2(nextOffset, 8);
>             memberQualifier.layoutXfbOffset = nextOffset;
>         } else
>             nextOffset = memberQualifier.layoutXfbOffset;

**Comment by infapi00** [comment]
> Yes. From GLSL 4.60 spec, section 4.4.2 "Output Layout Qualifiers", subsection "Transform Feedback Layout Qualifiers", on the "For example" table:

I  have just sent pull request #1567. It is the same fix I sent some weeks ago, but I updated one of the test cases, in order to use this one from GLSL spec.
